### PR TITLE
avx2: added implmentation for mm256_abs_x functions

### PR DESCRIPTION
**Pull request by masterchef2209**
avx2: added implmentation for mm256_abs_epi8, mm256_abs_epi16, mm256_abs_epi32

### CODE REVIEW

**Comment by masterchef2209** [comment]
Hi, I see there are many functions which are present in Intel Intrinsics Guide but not in issue tracker lists, do we not need implementation of those functions?

**Comment by masterchef2209** [comment]
I have used Method 2 from here for implementing abs
https://www.geeksforgeeks.org/compute-the-integer-absolute-value-abs-without-branching/

**Comment by mr-c** [comment]
> Hi, I see there are many functions which are present in Intel Intrinsics Guide but not in issue tracker lists, do we not need implementation of those functions?

Interesting! Can you be more specific?

**Comment by masterchef2209** [comment]
> > Hi, I see there are many functions which are present in Intel Intrinsics Guide but not in issue tracker lists, do we not need implementation of those functions?
> 
> Interesting! Can you be more specific?

Sorry, that was an error on my part, I misunderstood. Please ignore that comment.

**Comment by nemequ** [comment]
> I have used Method 2 from [here](https://www.geeksforgeeks.org/compute-the-integer-absolute-value-abs-without-branching/) for implementing abs
> I hope this will be faster than the usual check using ternary operator.

I'm really glad you did this, it's a great example.

It's actually [the same](https://godbolt.org/z/nGA-d2) if you're targeting the default ISA extensions (for x86_64, SSE2).  As in they actually generate the same code.  Both compilers recognize that the ternary-operator version is trying to compute an absolute value, so they output the fastest code they know about to do that.

However, things get more interesting if you actually enable AVX2 (but still force them to use the portable fallbacks instead of the AVX2 implementation in SIMDe).  In this case, the ternary operator version is [faster on GCC](https://godbolt.org/z/UBPXgy) since the compiler will actually generate a vpabsd instruction (which is what _mm256_abs_epi32 maps to).  On clang, [they're equal](https://godbolt.org/z/zj_o89) because clang actually recognizes *both* patterns as computing the absolute value and will generate optimal code for both.

Another thing to keep in mind is that using the ternary operator in SIMDe (or any code the compiler can autovectorize) doesn't typically result in a branch.  Think about how the comparison functions work; they return a mask with all bits set if the comparison is true, and all bits set if the comparison is false.  That means that the compiler can use simple bitwise operations to generate a result instead of branching.  Specifically, take a look at the [`_mm256_andnot_si256`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_andnot_si256&expand=33,339), [`_mm256_or_si256`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_or_si256&expand=33,339,301,339,774,4071), and [`_mm256_and_si256`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm256_and_si256&expand=33,339,301) functions.  They're extremely fast (1 cycle / 0.33 latency / 2 uops).  Comparison functions are [fast](https://uops.info/html-instr/VCMPPD_XMM_XMM_XMM_I8.html), too.

You can code the ternary operator as

```
__m256i m256i_ternary(__m256i cond, __m256i a, __m256i b) {
  return _mm256_or_si256(
      _mm256_and_si256(a, cond),
      _mm256_andnot_si256(b, cond));
}
```

Just pass the result of the comparison function as `cond`.  Of course, you don't really need to do this; compilers are all really good at autovectorizing this for you.  But hopefully it gives you a good idea of why the ternary operator isn't really something to be avoided for SIMD.

**Comment by masterchef2209** [comment]
Thank You, that was very informative, but I don't understand why travis build failed?
I will add commit using ternary operator soon.

File: simde/x86/avx2.h [hunk 1]
```diff
@@ -67,6 +67,84 @@ SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
...
+#else
+  simde__m256i_private
+    r_,
+    a_ = simde__m256i_to_private(a);
+
+#if defined(SIMDE_VECTOR_SUBSCRIPT_OPS)
+  r_.i32 = ( a_.i32 ^ (a_.i32 >> 31) ) - (a_.i32 >> 31);
+#else
+  SIMDE__VECTORIZE
+  for (size_t i = 0 ; i < (sizeof(r_.i8) / sizeof(r_.i8[0])) ; i++) {
```
[the remark this answers is not in the document]
Comment (masterchef2209) [reply simde/x86/avx2.h:136]: sorry i missed it, I should probably change my ide theme

**Comment by nemequ** [comment]
It looks like ICC is still a problem.  It might be a compiler bug.  Let's see if the problem persists after switching to the ternary operator.

**Comment by masterchef2209** [comment]
> It looks like ICC is still a problem. It might be a compiler bug. Let's see if the problem persists after switching to the ternary operator.

Yes, I am working on ternary operator commits. I hope it works.

**Comment by nemequ** [comment]
Looks great.  Merged, thanks!

Sorry about the delay.

**Pull request closed without merging**